### PR TITLE
Rewrite and complete j.l.String{Builder,Buffer}.

### LINKED ISSUE #3124: StringBuilder.delete(start:Int, end:Int) not implemented
It seems that the function delete(start:Int, end:Int) is not implemented.

### PR DESCRIPTION
**Pull request by sjrd**
This is a complete rewrite of `java.lang.StringBuilder`, with the complete set of methods from JDK 8. This includes `delete()`, which fixes #3124. The test suite has also been rewritten.

`java.lang.StringBuffer` is also rewritten, but delegates all its operations to an underlying `StringBuilder`. Since this class is rarely used, due to its bad reputation (synchronization), it is not worth making it fast.

A few methods of `java.lang.String` are also added, if they were necessary to implement the corresponding ones in `StringBuilder`. No effort was made to complete the set of methods of `String` in this commit.

### CODE REVIEW

**Comment by sjrd** [comment]
Dear Jenkins, retest this please

**Comment by sjrd** [comment]
Jenkins, retest this please. PhantomJS crashes more and more often these days :-(

**Comment by sjrd** [comment]
Hum, actually, the failure seems to be consistently on 2.10.2, and with no other version. Is it because 2.10.2 is slightly larger in terms of .js size, and we are crossing the magical barrier of doom for PhantomJS again? We already had to [compromise on testing with PhantomJS](https://github.com/scala-js/scala-js/blob/72456873a56f1ea833be996bb57707f2faecf52a/project/Build.scala#L1591-L1620). Do we need to cut another part of the test suite?

File: javalanglib/src/main/scala/java/lang/StringBuffer.scala [hunk 1]
```diff
@@ -1,177 +1,166 @@
...
+  def this() = this(new StringBuilder())
+  def this(str: String) = this(new StringBuilder(str))
+  def this(capacity: Int) = this(new StringBuilder(capacity))
+  def this(seq: CharSequence) = this(seq.toString)
 
-  def append(str: Array[scala.Char]): StringBuffer =
-    append(str, 0, str.length)
+  /** A helper so that we can write stuff like `withThisResult(append(obj))`. */
+  @inline
+  private def withThisResult[A](op: StringBuilder): StringBuffer = this
```
Comment (gzm0) [inline javalanglib/src/main/scala/java/lang/StringBuffer.scala]: Type parameter `A` is unused.

File: javalanglib/src/main/scala/java/lang/StringBuilder.scala [hunk 2]
```diff
@@ -1,59 +1,132 @@
...
-  def append(str: Array[scala.Char], offset: Int, len: Int): StringBuilder = {
-    var i = 0
-    while (i < len) {
-      content += str(i + offset)
-      i += 1
-    }
+  @inline
+  def append(obj: AnyRef): StringBuilder = {
+    // If (obj eq null), this appends "null", otherwise obj.toString()
+    content = content + obj
```
Comment (gzm0) [inline javalanglib/src/main/scala/java/lang/StringBuilder.scala]: `content += obj`?

File: javalanglib/src/main/scala/java/lang/StringBuilder.scala [hunk 3]
```diff
@@ -65,132 +138,91 @@ class StringBuilder(private var content: String) extends CharSequence
...
-      throw new StringIndexOutOfBoundsException(
-          s"Illegal to replace substring at [$start - $end] in string of length $length")
-    }
+  def capacity(): Int = length()
 
-    val realEnd = if (end > length) length else end // java api convention
-    content = content.substring(0, start) + str + content.substring(realEnd)
-    this
+  def ensureCapacity(minimumCapacity: Int): Unit = {
+    // Do nothing
```
Comment (gzm0) [inline javalanglib/src/main/scala/java/lang/StringBuilder.scala]: I would just write `def ensureCapacity(...): Unit = ()`. The comment "Do nothing" is not useful. If you want to document better, say **why** we are not doing anything. (But I think it is quite clear).

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/MockCharSequence.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js Test Suite        **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2017, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+package org.scalajs.testsuite.javalib.lang
+
+object MockCharSequence {
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/MockCharSequence.scala]: Call this a Fake... or a Dummy... or even WrappedStringCharSequence. A mock is something that allows to check if methods have been called.
See: https://stackoverflow.com/questions/346372/whats-the-difference-between-faking-mocking-and-stubbing

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+
+    assertEquals("null", resultFor(null))
+    assertEquals("hello", resultFor("hello"))
+  }
+
+  @Test def appendStringBuffer(): Unit = {
+    def resultFor(x: StringBuffer): String = newBuilder.append(x).toString()
+
+    assertEquals("null", resultFor(null))
+    assertEquals("hello", resultFor(new StringBuffer("hello")))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Add

```scala
assertEquals("", resultFor(new StringBuffer()))
```

?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 6]
```diff
@@ -0,0 +1,609 @@
...
+    val arr = Array('h', 'e', 'l', 'l', 'o')
+    assertEquals("hello", resultFor(arr, 0, 5))
+    assertEquals("ell", resultFor(arr, 1, 3))
+
+    if (executingInJVM)
+      expectThrows(classOf[NullPointerException], resultFor(null, 0, 0))
+
+    expectThrows(classOf[IndexOutOfBoundsException], resultFor(arr, -1, 2))
+    expectThrows(classOf[IndexOutOfBoundsException], resultFor(arr, 3, 3))
+    expectThrows(classOf[IndexOutOfBoundsException], resultFor(arr, 3, -2))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Add

```scala
expectThrows(classOf[IndexOutOfBoundsException], resultFor(arr, 0, 6))
```
?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 7]
```diff
@@ -0,0 +1,609 @@
...
+
+    val arr = Array('a', 'b', 'c', 'd', 'e')
+
+    assertEquals("0abcde12", resultFor("012", 1, arr))
+    assertEquals("ababcdeef", resultFor("abef", 2, arr))
+
+    expectThrows(classOf[StringIndexOutOfBoundsException],
+        resultFor("1234", -1, arr))
+    expectThrows(classOf[StringIndexOutOfBoundsException],
+        resultFor("1234", 6, arr))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Test the `null` case?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 8]
```diff
@@ -0,0 +1,609 @@
...
+  }
+
+  @Test def indexOfStringInt(): Unit = {
+    def resultFor(input: String, str: String, fromIndex: Int): Int =
+      initBuilder(input).indexOf(str, fromIndex)
+
+    assertEquals(7, resultFor("ababcdeabcf", "abc", 4))
+    assertEquals(2, resultFor("ababcdeabcf", "abc", 2))
+    assertEquals(-1, resultFor("ababcdeabcf", "abc", 10))
+    assertEquals(-1, resultFor("ababcdeabcf", "acb", 2))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Test the case `fromIndex >= input.length`?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 9]
```diff
@@ -0,0 +1,609 @@
...
+  }
+
+  @Test def lastIndexOfStringInt(): Unit = {
+    def resultFor(input: String, str: String, fromIndex: Int): Int =
+      initBuilder(input).lastIndexOf(str, fromIndex)
+
+    assertEquals(2, resultFor("ababcdeabcf", "abc", 2))
+    assertEquals(2, resultFor("ababcdeabcf", "abc", 6))
+    assertEquals(-1, resultFor("ababcdeabcf", "abc", 1))
+    assertEquals(-1, resultFor("ababcdeabcf", "acb", 10))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Test the case `fromIndex >= input.length`?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 10]
```diff
@@ -0,0 +1,609 @@
...
+    assertEquals("987654321", resultFor("123456789"))
+    assertEquals("dc\ud801\udc02ba", resultFor("ab\ud801\udc02cd"))
+    assertEquals("dc\ud802\ud801ba", resultFor("ab\ud801\ud802cd"))
+    assertEquals("dc\udc02\udc01ba", resultFor("ab\udc01\udc02cd"))
+    assertEquals("\ud801ba", resultFor("ab\ud801"))
+    assertEquals("dc\udc02", resultFor("\udc02cd"))
+  }
+
+  @Test def length(): Unit = {
+    assertEquals(5, initBuilder("hello").length())
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Test with surrogates?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 11]
```diff
@@ -0,0 +1,609 @@
...
+  }
+
+  @Test def capacity(): Unit = {
+    assertTrue(initBuilder("hello").capacity() >= 5)
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Test with surrogates?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 12]
```diff
@@ -0,0 +1,609 @@
...
+    assertEquals("foo", b.toString)
+    b.setLength(6)
+    assertEquals("foo\u0000\u0000\u0000", b.toString)
+  }
+
+  @Test def charAt(): Unit = {
+    def resultFor(input: String, index: Int): Char =
+      initBuilder(input).charAt(index)
+
+    assertEquals('e', resultFor("hello", 1))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Test with surrogates? (just maybe...)

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala [hunk 13]
```diff
@@ -0,0 +1,609 @@
...
+      initBuilder(input).codePointAt(index)
+
+    assertEquals(0x61, resultFor("abc\ud834\udf06def", 0))
+    assertEquals(0x1d306, resultFor("abc\ud834\udf06def", 3))
+    assertEquals(0xdf06, resultFor("abc\ud834\udf06def", 4))
+    assertEquals(0x64, resultFor("abc\ud834\udf06def", 5))
+    assertEquals(0x1d306, resultFor("\ud834\udf06def", 0))
+    assertEquals(0xdf06, resultFor("\ud834\udf06def", 1))
+    assertEquals(0xd834, resultFor("\ud834abc", 0))
+    assertEquals(0xdf06, resultFor("\udf06abc", 0))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringBuilderTest.scala]: Test out of bound (n.b. with surrogates)? (same on next)

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala [hunk 14]
```diff
@@ -142,6 +142,17 @@ class StringTest {
...
 
+  @Test def codePointBefore(): Unit = {
+    assertEquals(0x61, "abc\ud834\udf06def".codePointBefore(1))
+    assertEquals(0x1d306, "abc\ud834\udf06def".codePointBefore(5))
+    assertEquals(0xd834, "abc\ud834\udf06def".codePointBefore(4))
+    assertEquals(0x64, "abc\ud834\udf06def".codePointBefore(6))
+    assertEquals(0x1d306, "\ud834\udf06def".codePointBefore(2))
+    assertEquals(0xd834, "\ud834\udf06def".codePointBefore(1))
+    assertEquals(0xd834, "\ud834abc".codePointBefore(1))
+    assertEquals(0xdf06, "\udf06abc".codePointBefore(1))
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StringTest.scala]: Test out of bound?

**Comment by sjrd** [comment]
Rebased to benefit from #3138.

**Pull request merged**